substrate/frame/transaction-payment/asset-conversion-tx-payment/src/payment.rs

```diff
@@ -49,150 +53,219 @@ pub trait OnChargeAssetTransaction<T: Config> {
+				return Err(InvalidTransaction::Payment.into())
+			},
+		};
 
-		// charge the fee in native currency
-		<T::OnChargeTransaction>::withdraw_fee(who, call, info, fee, tip)
-			.map(|r| (r, native_asset_required, asset_consumed.into()))
+		// Since the exact price for `fee` has been quoted, the change should be zero.
+		ensure!(change.peek().is_zero(), InvalidTransaction::Payment);
+
+		Ok((fee_credit, asset_fee))
 	}
 
-	/// Correct the fee and swap the refund back to asset.
-	///
-	/// Note: The `corrected_fee` already includes the `tip`.
-	/// Note: Is the ED wasn't needed, the `received_exchanged` will be equal to `fee_paid`, or
-	/// `fee_paid + ed` otherwise.
 	fn correct_and_deposit_fee(
 		who: &T::AccountId,
-		dispatch_info: &DispatchInfoOf<T::RuntimeCall>,
-		post_info: &PostDispatchInfoOf<T::RuntimeCall>,
-		corrected_fee: BalanceOf<T>,
-		tip: BalanceOf<T>,
-		fee_paid: LiquidityInfoOf<T>,
-		received_exchanged: Self::LiquidityInfo,
+		_dispatch_info: &DispatchInfoOf<<T>::RuntimeCall>,
+		_post_info: &PostDispatchInfoOf<<T>::RuntimeCall>,
+		corrected_fee: Self::Balance,
+		tip: Self::Balance,
 		asset_id: Self::AssetId,
-		initial_asset_consumed: AssetBalanceOf<T>,
-	) -> Result<AssetBalanceOf<T>, TransactionValidityError> {
-		// Refund the native asset to the account that paid the fees (`who`).
-		// The `who` account will receive the "fee_paid - corrected_fee" refund.
-		<T::OnChargeTransaction>::correct_and_deposit_fee(
-			who,
-			dispatch_info,
-			post_info,
-			corrected_fee,
-			tip,
-			fee_paid,
-		)?;
-
-		// calculate the refund in native asset, to swap back to the desired `asset_id`
-		let swap_back = received_exchanged.saturating_sub(corrected_fee);
-		let mut asset_refund = Zero::zero();
-		if !swap_back.is_zero() {
-			// If this fails, the account might have dropped below the existential balance or there
-			// is not enough liquidity left in the pool. In that case we don't throw an error and
-			// the account will keep the native currency.
-			match CON::swap_exact_tokens_for_tokens(
-				who.clone(), // we already deposited the native to `who`
-				vec![
-					N::get(),        // we provide the native
-					asset_id.into(), // we want asset_id back
-				],
-				swap_back,   // amount of the native asset to convert to `asset_id`
-				None,        // no minimum amount back
-				who.clone(), // we will refund to `who`
-				false,       // no need to keep alive
+		already_withdrawn: Self::LiquidityInfo,
+	) -> Result<BalanceOf<T>, TransactionValidityError> {
+		let (fee_paid, initial_asset_consumed) = already_withdrawn;
+		let refund_amount = fee_paid.peek().saturating_sub(corrected_fee);
+		let (fee_in_asset, adjusted_paid) = if refund_amount.is_zero() ||
+			F::total_balance(asset_id.clone(), who).is_zero()
+		{
+			// Nothing to refund or the account was removed be the dispatched function.
+			(initial_asset_consumed, fee_paid)
```

Comment:
right. also current implementation for only a native asset (pallet-transaction-payment) also follows this logic and asserts it with a test.  there might be other reasons. for example if you transfer all and wanna your account to be removed, you do not wanna any refund back.
may be @kianenigma can add more to this.

Reply:
@georgepisaltu which assumption? the transfer all and reap account is real use case.

Reply:
The assumption that the account was removed. I understand the reap account is a real use case, I was talking about @ggwpez 's situation where you maybe consumed all the asset. It would get messy with the ED calculation, but I think we don't care about that if the asset is sufficient and we have some of it to refund, right? I'm asking if the intended behavior is to ignore this "sufficient" case as with the current impl.

Reply:
Yes, system account or any other can be still alive.
The intended behavior is to have no refund if a dispatched call results the asset account to be removed, regardless of the asset type.